napari/utils/context/_tests/test_expressions.py

```diff
@@ -202,3 +202,18 @@ def test_serdes(expr):
 def test_bad_serdes(expr):
     with pytest.raises(SyntaxError):
         parse_expression(expr)
+
+
+def test_eval_expression():
+    """Test that expression.eval works with ContextKeyServices"""
+    from napari.utils.context._service import ContextKeyService
+
+    expr = parse_expression('a and b > 5')
+    context = ContextKeyService()
+    context['a'] = False
+    context['b'] = 8
+    assert expr.eval(context) is False
+    context['a'] = True
+    assert expr.eval(context) is True
+    context['b'] = 3
+    assert expr.eval(context) is False
```

Comment:
Jut to make sure I understand, 

Here you are evaluating in the root context, but you should be able to evaluate in the any scope context ? Is that right ?  Do you want actually evaluate in a scoped context ? 

I'm not too sure I see yet how this will interact with multiple objects and `when` clauses, in particular when you have many instances for examples of layers, but I guess you know where you are going, and in particular why you `root.get_context(obj)`, and not `root.get_context(key)`, I I would have thought that in expressions like `layer.active.n_dims > 2` for example, there would be a `root.get_context('layer').get_context('active').get_context('n_dims')`. But i'm likely misunderstanding how this will be used.

Reply:
> Here you are evaluating in the root context, but you should be able to evaluate in the any scope context ? Is that right ? Do you want actually evaluate in a scoped context ?

yes that's correct.  for the purpose of this test, I'm mostly making sure that I can evaluate differently.  normally we'd be evaluating in a scoped context I assume.

> But i'm likely misunderstanding how this will be used.

No, I think your question is a really good one.  For the moment, I was just focusing on the concept of a scoped context. but _exactly_ how we pick a context to evaluate is still somewhat of a question in my mind.  I guess, where a viewer has a number of singleton objects (like a single layerlist for instance), we could easily merge specific contexts prior to evaluation.  but that is indeed an open question